### DISCUSSION TITLE
Variadic functions not accessible through introspected bindings

### DISCUSSION THREAD

**Discussion by jtojnar** [libvips/libvips#3251] [Ideas]
I wanted to use `vips_autorot` function in a Haskell program but it looks like the bindings do not contain anything other than the [`vips_autorot_remove_angle`](https://hackage.haskell.org/package/gi-vips-8.0.3/docs/GI-Vips-Objects-Image.html#v:imageAutorotRemoveAngle) helper.

Looking at the 

https://github.com/libvips/libvips/blob/feae09e9cdc06a6667bc209d655f9c4bc78ddb4a/libvips/conversion/autorot.c#L256

it is actually a variadic function so it is not really bindable without manual overrides. Would it be feasible to add bindable functions similar to the ones provided by [GSettings API](https://docs.gtk.org/gio/class.Settings.html#constructors):

```c
int vips_autorot_with_flip( VipsImage *in, VipsImage **out, bool flip );
int vips_autorot_with_angle( VipsImage *in, VipsImage **out, VipsAngle angle );
int vips_autorot_full( VipsImage *in, VipsImage **out, VipsAngle angle, bool flip );
```

**Comment by jcupitt** [comment]
Hi @jtojnar,

The `autorot` operator should be there. I see:

```
 $ vips -l | grep auto
      VipsAutorot (autorot), autorotate image by exif tag
```

Are you trying to use libvips via gobject-introspection? There's a chapter in the libvips docs about binding:

https://www.libvips.org/API/current/binding.html

Summary: 

- libvips has two APIs: 
  1. the usual C API with `vips_add()` etc., which is varadic, and 
  2. a lower-level GObject API (no varargs) designed for interfacing 
- API 1) is a very thin layer over API 2) and just there for the convenience of human C programmers. 
- Other languages should use API 2), and avoid 1)
- All the bindings we've made use this pattern
- You can call into API 2) with gobject-introspection, you don't need to write any C code

**Reply by jtojnar** [reply]
Ah, I skipped that page since I was using an existing binding library instead of creating a new one. And yes, the existing [Haskell bindings](https://hackage.haskell.org/package/gi-vips) are based on g-i.

Your explanation makes sense, though it is not very ergonomic.

But it appears there is a [more higher-level Haskell library](https://github.com/xprl-gjf/hvips) so I will either use that or just execute `vipsthumbnail` process.
